Dask data manager

bjlittle opened the pull request:
Introduces the concept of a generic data manager with a well defined API and strictly enforced contract. 

The purpose, and intent, of the data manager is to abstract and contain the state and behaviour required to manage generic iris data payload, be it for a cube, coordinate points, coordinate bounds, cell measure payload etc etc

This is a WIP at the moment ...

- [x] Complete `XXX` doc-strings
- [x]  Unit test coverage for `iris._data_manager.DataManager._deepcopy` method
- [x] Remove `lib/iris/manager.txt` file

bjlittle commented:
Closes #2386

lbdreyer commented on lib/iris/_data_manager.py:
```diff
@@ -0,0 +1,392 @@
+        .. note::
+            Any lazy data being managed will be realised.
+
+        """
```
I think this doc string need rewording. Usually the Returns section would have the data types.
Perhaps:
```
Returns the real data. If any lazy data is being managed, it will be realised.

Returns:
    the real :class:`~numpy.ndarray` or   :class:`numpy.ma.core.MaskedArray`

```

lbdreyer commented on lib/iris/manager.txt:
```diff
@@ -0,0 +1,94 @@
+_lazy_data.py
```
Presumably this text file will be removed at some point. Or at least turned into some documentation??

bjlittle commented on lib/iris/_data_manager.py:
```diff
@@ -0,0 +1,392 @@
+    def _assert_axioms(self):
+        """
+        Definition of the manager state, that should never be violated.
+
```
Personally, I think it's a bit much. It's a private method, so it's only there for the developer, and they can just read the code - so I'm seeing the doc-string giving some context to the dev not a literal translation of what it actually does.

bjlittle commented on lib/iris/_data_manager.py:
```diff
@@ -0,0 +1,392 @@
+        """
+        # Ensure there is a valid data state.
+        is_lazy = bool(self._lazy_array is not None)
+        is_real = bool(self._real_array is not None)
```
Yup, good spot ... this was originally actually something else and needed the `bool`

bjlittle commented on lib/iris/_data_manager.py:
```diff
@@ -0,0 +1,392 @@
+            else:
+                # Check that the replacement data is valid relative to
+                # the currently managed data.
+                DataManager(self.core_data).replace(data)
```
Yeah, so I found  this all a bit wierd ...

We're doing a copy, but giving convenience to allow the `data` to be changed (which in my head is almost another way of creating a different `DataManager`).

But when different `data` is provided we can't just use it as provided, the `data` must be relative to the `data` being copied ... this line was the neatest way I could find to check that the `data` replacement is valid i.e. no cheeky changing of shape relative to the original `data`.

I could do
```python
result = DataManager(self.core_data).replace(data, realised_dtype=self._realised_dtype)
```
instead, but I'd also need to create the `result` for the case when no `data` is provided ... I'll see what that's like ... might be less wierd

bjlittle replied:
Actually, I stand-by the original, as it is a clean separation between checking the `data` replacement and `realised_dtype` replacement

lbdreyer replied:
Sorry! This was actually a note to myself that I had meant to delete, before I submitted the review. I do get the purpose of this

>Actually, I stand-by the original, as it is a clean separation between checking the data replacement and realised_dtype replacement ... convince me otherwise :wink:

Any risk of memory problems? If you are working with big real data that you replace with more big data, you are creating an extra DataManager that is only a check but doesn't get used. Perhaps delete it?

bjlittle replied:
But, if I'm thinking about this right, `self.core_data` already exists, and `data` already exists (as the owner passed it in) ... so this operation is not duplicating or increasing the process memory footprint by the same relative scale. I think this is safe, even for big data.

lbdreyer replied:
Good point. I think I had mistakenly thought that you were taking a copy of `data`.

bjlittle commented on lib/iris/_data_manager.py:
```diff
@@ -0,0 +1,392 @@
+
+        return result
+
+    def _dtype_setter(self, realised_dtype):
```
`dtype` is right, as it returns the `dtype` of the data OR the `realised_dtype` ... that mirrors what we currently have in the `dask` branch, hence the use of `dtype_setter`. On reflection, I'm happy to change this to `_realised_dtype_setter`

lbdreyer replied:
I do prefer `_realised_dtype_setter`

bjlittle commented on lib/iris/_data_manager.py:
```diff
@@ -0,0 +1,392 @@
+        """
+        return self._real_array is not None
+
+    def lazy_data(self):
```
The owner of the `DataManager` will call this. The use case is all over the `dask` branch wrt the `Cube`. 

Eventually, this will also be the case for coordinates and the like ...

bjlittle replied:
Nope. I'm attempting to mirror the current decisions (rightly or wrongly) in the `dask` branch based on the `Cube` usage. If we need to have that discussion (along with renaming `core_data` and how to access the `data` etc) then I think (for my sanity) it should take place in a separate PR.

Does that sound reasonable?

lbdreyer replied:
Makes sense. I, admittedly, had up to this point only been thinking of this as an independent thing.

bjlittle replied:
Yeah, the behaviour has really been driven by the owner i.e. the usage pattern in the `Cube` so I'm hoping that once this is merged, the act of integration will prove what we should (and should not) keep in the `DataManager` API

lbdreyer replied:
Hmmm ... I can't say I like it.... (my worry is someone tries to use it to add a lazy array) but I'm not overly fussed so long as it get addressed in an upcoming PR.

bjlittle commented on lib/iris/_data_manager.py:
```diff
@@ -0,0 +1,392 @@
+        The number of dimensions covered by the data being managed.
+
+        """
+        return len(self.shape)
```
Yup, got lazy and mirrored the `Cube.ndim` method ... *sheesh*

bjlittle commented on lib/iris/tests/unit/data_manager/test_DataManager.py:
```diff
@@ -0,0 +1,602 @@
+        result = dm.lazy_data()
+        self.assertTrue(dm.has_real_data())
+        self.assertEqual(result, self.lazy_array)
+        self.assertTrue(dm.has_real_data())
```
How would that matter? I need to draw the line at some point and trust that "stuff just works". So my argument is that the unit test coverage for `iris._lazy_data.as_lazy_data` does that - I really don't want to replicate that here.

bjlittle commented on lib/iris/tests/unit/data_manager/test_DataManager.py:
```diff
@@ -0,0 +1,602 @@
+        shape = (2, 3, 4)
+        size = np.cumprod(shape)[-1]
+        self.real_array = np.arange(size).reshape(shape)
+        self.lazy_array = da.from_array(self.real_array, chunks=shape)
```
Yeah, I hummed and hawed about this ... but was more focused on churning out as much test coverage as possible.

I think your right and I should rely on the goodness of `as_lazy_data` throughout

bjlittle commented on lib/iris/tests/unit/data_manager/test_DataManager.py:
```diff
@@ -0,0 +1,602 @@
+        self.assertIsInstance(result, ma.MaskedArray)
+        self.assertIsNone(dm._realised_dtype)
+        self.assertEqual(dm.dtype, self.realised_dtype)
+        self.assertArrayEqual(result, self.lazy_mask_array_masked)
```
It's not. The testing framework is realizing the `self.lazy_mask_array_masked` under the hood.

I'll make it more explicit i.e. `self.lazy_mask_array_masked.compute()`

bjlittle commented on lib/iris/tests/unit/data_manager/test_DataManager.py:
```diff
@@ -0,0 +1,602 @@
+        dm.data = value
+        self.assertTrue(dm.has_real_data())
+        self.assertFalse(dm.has_lazy_data())
+        self.assertArrayEqual(dm.data, np.array(value))
```
I could but it makes me do more work, as the `DataManager` data might be lazy, so `core_data` would return a lazy thing, that I then need to realise. Using the `DataManager.data` getter does all that for me.

lbdreyer replied:
Yeah I had made this comment before I read the rest of your tests. I thought it would be a good thing cause then you're being consistent regardless of whether it returns a lazy array or real array but that doesn't really fit what you're testing 
I'm happy with how you've done ti.

bjlittle commented on lib/iris/tests/unit/data_manager/test_DataManager.py:
```diff
@@ -0,0 +1,602 @@
+        size = np.cumprod(shape)[-1]
+        mask_array = ma.arange(size).reshape(shape)
+        dtype = mask_array.dtype
+        lazy_array = as_lazy_data(mask_array)
```
I've addressed this now.

bjlittle commented on lib/iris/tests/unit/data_manager/test_DataManager.py:
```diff
@@ -0,0 +1,602 @@
+        self.assertFalse(dm.has_lazy_data())
+        self.assertArrayEqual(dm.data, np.array(value))
+
+    def test_zero_ndim_real_with_scalar_float(self):
```
I've not checked unsigned or bool .... where does the testing stop?

lbdreyer replied:
> where does the testing stop?

It never stops :weary:  :sob:

But yeah, I'm happy with the testing as it is.

bjlittle commented on lib/iris/tests/unit/data_manager/test_DataManager.py:
```diff
@@ -0,0 +1,602 @@
+
+    def test_coerce_to_ndarray(self):
+        shape = (2, 3)
+        size = np.cumprod(shape)[-1]
```
If I used them in-place, then I'd be repeating the `(2, 3)` or using the `(2, 3)` and a hardwired size of `6` that is linked to the shape, opening up the possibility of a dev changing one and not the other. It's just my personal choice ... but if you feel strongly enough about it, then I'm happy to take them out.

lbdreyer replied:
I'm happy with it as it is (it also maintains consistency with how the other tests are written), it was just something I had noticed.

bjlittle commented on lib/iris/tests/unit/data_manager/test_DataManager.py:
```diff
@@ -0,0 +1,602 @@
+
+#
+# TBD: tests for _deepcopy,
+#
```
I don't think we need explicit tests for `__init__`, otherwise I feel that I would be repeating an awful lot of the current tests of the individual methods. I'm taking comfort in the fact that I've just written 61 unit tests ... plus there is more to come, and the behaviour and state of the `__init__` is pretty inline with outsourcing to the class methods ...

lbdreyer replied:
Yeah, I agree with not testing __init__ :+1:

bjlittle commented on lib/iris/tests/unit/data_manager/test_DataManager.py:
```diff
@@ -0,0 +1,602 @@
+            mocker.return_value = return_value
+            result = copy.deepcopy(dm)
+            self.assertEqual(mocker.call_count, 1)
+        self.assertIs(result, return_value)
```
I'll take a peek ...

bjlittle replied:
Yeah, using `copy.deepcopy` populates the `memo` (somehow) that's why I dodged that bullet ... I'll see if I can work out what it's doing and extend the test ... seems do-able!

bjlittle replied:
Hmmm this is actually become some what painful ....

For this test `copy.deepcopy` populates the `memo` with the following example:
```python
   (Pdb) list
556  	        return_value = mock.sentinel.return_value
557  	        with mock.patch(method) as mocker:
558  	            mocker.return_value = return_value
559  	            result = copy.deepcopy(dm)
560  	            import pdb; pdb.set_trace()
561  ->	            self.assertEqual(mocker.call_count, 1)
562  	        self.assertIs(result, return_value)
563  	
564  	
565  	class Test_copy(tests.IrisTest):
566  	    def test(self):
(Pdb) p mocker.call_args
call({139777040548944: sentinel.return_value, 139777048606992: [DataManager(array(0))]})
```
If we want to introspect at this level, then I'm pretty much forced to implement `__eq__` and `__ne__` comparison methods on the `DataManager` ... that might be nice to have, but I'm not really motivated to do so just for a test. It may be that there is a genuine use case, say in `concatenate` or `merge` or somewhere else that might would be super handy, but I'm erring on the side of not doing it here and now .... thoughts @lbdreyer ?

lbdreyer replied:
Interesting that we don't have tests for cube._deepcopy 
Yep, I'm happy for this to be left for someother time. We don't gain much from doing it and I had suspected it would be quite painful. So in the name of diminishing returns I would agree to putting this off.

bjlittle commented on lib/iris/_data_manager.py:
```diff
@@ -0,0 +1,392 @@
+            Boolean.
+
+        """
+        return self._lazy_array is not None
```
My expectation is that it's a convenience for the owner ... certainly that's what I've seen in terms of usage i.e. `if not self.has_lazy_data()` rather than `if self.has_real_data()` or just flip the conditional actions ... either way, I'm pretty undecided and I'm the one suggesting it - we could certainly live without it as `has_lazy_data()` is the minimum that we might need ... I'm just floating the idea and relying on the wisdom of the crowd :wink:

dkillick replied:
I'm a fan of `has_real_data`:

* as a reader of code it is more clear what's being checked than `not has_lazy_data`, which is more than twice as hard to process as you've got to work out the whole statement from the combination of the two.
* as a user of Iris it's nice to be able to check exactly what I want rather than the inverse of what I don't want to check.
* Lines of code are cheap! We shouldn't be withholding convenience functions for the sake of fewer lines of source code...

bjlittle replied:
@dkillick I pre-empted your thoughts and nuked it ... it is now but a whim floating away in a Spring zephyr ...

bjlittle commented:
@lbdreyer Okay, I think that I've addressed all of your issues apart from the open questions:
- do we need the `has_real_data` method
- is the method `lazy_data` a property

I still need to update the `XXX` doc-strings, but all the other material changes, I think, are pretty much there ...

lbdreyer commented on lib/iris/_data_manager.py:
```diff
@@ -76,12 +79,69 @@ def __deepcopy__(self, memo):
+
+        if isinstance(other, type(self)):
+            result = False
+            is_lazy = self.has_lazy_data() == other.has_lazy_data()
```
Trouble with this naming is that it suggests that, if `is_lazy` is True that both self and other have lazy data. But the following could be the case:
```
>>> print self.has_lazy_data()
False
>>> print other.has_lazy_data()
False
>>> is_lazy = self.has_lazy_data() == other.has_lazy_data()
>>> print is_lazy
True
```

What about `is_lazy_check` or `both_lazy`?

bjlittle replied:
I'll do a name change ... I've almost ran out of creative juice, so thanks for the suggestions :wink:

lbdreyer commented:
The ordering of the class methods, properties etc is different to the ordering of the tests.
In the class itself the order is:
```
__init__
__things with double leading underscore in alphabetical order
_thing with single leading underscore in alphabetical order
The rest in alphabetical order
```

Should the unit tests also follow this order?

bjlittle commented:
@lbdreyer I think this is good to merge now ...

pelson commented:
I have no comment on the implementation, but I want to explore whether this an abstraction too far.
The PR itself obviously only *enables* the data manager abstraction, rather than making use of the implementation within the Cube, so it isn't clear to us as reviewers how much benefit we are able to get out of this change. Personally, I'd rather *reduce* our layers of abstraction within Iris and start treating arrays as ducks.

Would you be able to put forward a separate PR that makes use of the ```DataManager``` so that we can see how much code it liberates?

bjlittle commented:
@pelson That's the follow on PR ...

pelson commented:
> @pelson That's the follow on PR ...

:smile: - and I appreciate you separating the two PRs! But to add technical complexity without being able to measure that against the simplification it will bring means we are decisions in the dark.

pelson commented:
Just realised this is against the dask branch. If the intention is for that branch to be reviewed before moving to master, ignore my question about technical complexity - that can be answered in the larger review.

dkillick commented on lib/iris/_data_manager.py:
```diff
@@ -0,0 +1,471 @@
+            self._assert_axioms()
+
+    @property
+    def core_data(self):
```
As a note for our future selves, I still like the idea of calling this `data`. Then we can rename `data` to `real_data` and keep `lazy_data` as `lazy_data`. I think this is the most descriptive set of names for these variables.

But let's not do that in this PR! One thing at a time.

bjlittle replied:
Agreed. We should create a ticket in the backlog really.

bjlittle replied:
I've created a note in the `iris-dask-backlog` to cover this @dkillick

dkillick commented on lib/iris/_data_manager.py:
```diff
@@ -0,0 +1,471 @@
+        if not (hasattr(data, 'shape') and hasattr(data, 'dtype')):
+            data = np.asanyarray(data)
+
+        # Determine whether the class instance has been created,
```
It feels like there should be a better way to check for instantiation than this...

bjlittle replied:
Please, let me know ...

dkillick replied:
I wish I knew what it was...

lbdreyer commented on lib/iris/_data_manager.py:
```diff
@@ -0,0 +1,471 @@
+        Kwargs:
+
+        * realised_dtype:
+            The intended dtype of the specified lazy data.
```
I think it might be worth stating this should only be set for non-float data. Else someone may do:
`DataManager(lazy_data, realised_dtype=np.float32)`
which would return 
`Can only cast lazy data to an integer or boolean dtype, got float.`

bjlittle replied:
Done.

bjlittle commented:
@lbdreyer I've just rebased against `upstream/dask` ...

lbdreyer commented on lib/iris/tests/unit/data_manager/test_DataManager.py:
```diff
@@ -0,0 +1,775 @@
+                          realised_dtype=np.dtype('int16'))
+        self.assertFalse(dm1 == dm2)
+
+    def test__non_DataManager_failure(self):
```
why does this have 2 underscores, i.e. why is it not `test_non_DataManager_failure`

bjlittle replied:
It's the type of test, see the dev notes on testing name styles. Not fussed really.

lbdreyer commented:
Thanks @bjlittle and really well done for persevering and not losing your mind in the process :clap: :clap:  :100:   :relieved:
